File: JANL/ExtendedControls/DGVFilter.vb
Option Strict On

Imports System.Drawing

''' <summary>
''' Фильтр для DataGridView
''' </summary>
Public Class DGVFilter
	Inherits ToolStripTextBox
	Private WithEvents DGV As DataGridView
	Private WithEvents Timer As New Timer() With {.Interval = 1000}
	Private ReadOnly OffColor As Color = Color.Gray
	Private ReadOnly OffText As String = "Фильтр"
	Private Collumns As IEnumerable(Of String)

	Public Sub New()
		DefaultLanguage = InputLanguage.FromCulture(New Globalization.CultureInfo("ru-RU"))
	End Sub

	''' <summary>
	''' Инициализация
	''' </summary>
	''' <param name="DGV">Таблица для поиска</param>
	''' <param name="Collumns">Столбцы для поиска</param>
	Public Sub Init(DGV As DataGridView, Collumns As IEnumerable(Of String))
		Me.DGV = DGV
		Me.Collumns = Collumns
		With DGV.FindForm
			.KeyPreview = True
			AddHandler .KeyDown, AddressOf DGV_Search_F
		End With
		ForeColor = OffColor
		Text = OffText
	End Sub

	Public Sub SetFilter(str As String)
		ForeColor = If(String.IsNullOrWhiteSpace(str), OffColor, Color.Empty)
		Text = str
		ApplyFilter()
	End Sub

	Protected Overrides Function ProcessCmdkey(ByRef msg As Message, keyData As Keys) As Boolean
		Select Case keyData
			Case Keys.Escape
				Text = ""
				ApplyFilter()
				Return True
			Case Keys.Enter
				ApplyFilter()
				Return True
			Case Else
				Return MyBase.ProcessCmdKey(msg, keyData)
		End Select
	End Function

	Private Sub ApplyFilter()
		Timer.Stop()
		Dim BS = CType(DGV.DataSource, BindingSource)
		If Text = "" Then
			BS.Filter = Nothing
			Exit Sub
		End If
		Dim LF = Collumns.Select(Function(c) String.Format("({0} Like '%" + Text.Trim.Replace(" ", "%' AND {0} Like '%") + "%')", c))
		BS.Filter = String.Join(" OR ", LF)
		OnFilterApplied()
	End Sub

#Region "Properties"
	Public Property DefaultLanguage As InputLanguage
#End Region

#Region "UI Events"

	Private Sub DGV_Search_F(sender As Object, e As KeyEventArgs) Handles DGV.KeyDown
		If e.Control AndAlso e.KeyCode = Keys.F Then
			Focus()
			e.Handled = True
		End If
	End Sub

	Private Sub DGV_Search_GotFocus(sender As Object, e As EventArgs) Handles Me.GotFocus
		If ForeColor = OffColor Then
			ForeColor = Color.Empty
			Text = ""
		End If
		InputLanguage.CurrentInputLanguage = DefaultLanguage
	End Sub

	Private Sub DGV_Search_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
		Timer.Stop()
		Timer.Start()
	End Sub

	Private Sub DGV_Search_LostFocus(sender As Object, e As EventArgs) Handles Me.LostFocus
		If Text.Length = 0 Then
			ApplyFilter()
			ForeColor = OffColor
			Text = OffText
		End If
	End Sub

#End Region

	'Private Sub DGV_Search_KeyUp(sender As Object, e As KeyEventArgs) Handles Me.KeyUp
	'	If e.KeyCode = Keys.Back Then isBack = False  Filter Applied
	'End Sub
	Private Sub Timer_Tick(sender As Object, e As EventArgs) Handles Timer.Tick
		ApplyFilter()
	End Sub

#Region "Events"

	Protected Sub OnFilterApplied()
		RaiseEvent FilterApplied()
	End Sub

	Public Event FilterApplied()

#End Region

End Class

File: JANL.WinForms/ExtendedControls/DGVFilter.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace JANL
{
    /// <summary>
    /// Фильтр для DataGridView
    /// </summary>
    public class DGVFilter : ToolStripTextBox
    {
        private readonly Color OffColor = Color.Gray;
        private readonly string OffText = "Фильтр";
        private readonly Timer Timer = new Timer() { Interval = 1000 };
        private IEnumerable<string> Collumns;
        private DataGridView DGV;

        public DGVFilter()
        {
            DefaultLanguage = InputLanguage.FromCulture(new System.Globalization.CultureInfo("ru-RU"));
            Timer.Tick += Timer_Tick;
            LostFocus += DGVFilter_LostFocus;
            GotFocus += DGVFilter_GotFocus;
            KeyDown += DGVFilter_KeyDown;
        }

        /// <summary>
        /// Инициализация
        /// </summary>
        /// <param name="DGV">Таблица для поиска</param>
        /// <param name="Collumns">Столбцы для поиска</param>
        public void Init(DataGridView DGV, IEnumerable<string> Collumns)
        {
            this.DGV = DGV;
            this.Collumns = Collumns;
            {
                var Parent = DGV.FindForm();
                Parent.KeyPreview = true;
                Parent.KeyDown += DGV_Search_F;
            }
            ForeColor = OffColor;
            Text = OffText;
        }

        public void SetFilter(string str)
        {
            ForeColor = string.IsNullOrWhiteSpace(str) ? OffColor : Color.Empty;
            Text = str;
            ApplyFilter();
        }

        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
        {
            switch (keyData)
            {
                case Keys.Escape:
                    Text = "";
                    ApplyFilter();
                    return true;

                case Keys.Enter:
                    ApplyFilter();
                    return true;

                default:
                    return base.ProcessCmdKey(ref msg, keyData);
            }
        }

        private void ApplyFilter()
        {
            Timer.Stop();
            var BS = (BindingSource)DGV.DataSource;
            if (Text == "")
            {
                BS.Filter = null;
                return;
            }
            var LF = Collumns.Select(c => string.Format("({0} Like '%" + Text.Trim().Replace(" ", "%' AND {0} Like '%") + "%')", c));
            BS.Filter = string.Join(" OR ", LF);
            OnFilterApplied();
        }

        #region Properties

        public InputLanguage DefaultLanguage { get; set; }

        public int WaitTime
        {
            get => Timer.Interval;
            set => Timer.Interval = value;
        }

        #endregion Properties

        #region UI Events

        private void DGV_Search_F(object sender, KeyEventArgs e)
        {
            if (e.Control && e.KeyCode == Keys.F)
            {
                Focus();
                e.Handled = true;
            }
        }

        private void DGVFilter_GotFocus(object sender, EventArgs e)
        {
            if (ForeColor == OffColor)
            {
                ForeColor = Color.Empty;
                Text = "";
            }
            InputLanguage.CurrentInputLanguage = DefaultLanguage;
        }

        private void DGVFilter_KeyDown(object sender, KeyEventArgs e)
        {
            Timer.Stop();
            Timer.Start();
        }

        private void DGVFilter_LostFocus(object sender, EventArgs e)
        {
            if (Text.Length == 0)
            {
                ApplyFilter();
                ForeColor = OffColor;
                Text = OffText;
            }
        }

        #endregion UI Events

        private void Timer_Tick(object sender, EventArgs e) => ApplyFilter();

        #region Events

        protected void OnFilterApplied() => FilterApplied?.Invoke(this, EventArgs.Empty);

        public event EventHandler FilterApplied;

        #endregion Events
    }
}